### DISCUSSION TITLE
PatchOperation not updating values?

### DISCUSSION THREAD

**Discussion by KezimaDynamicsTK** [ravendb/ravendb#18801] [Q&A] [answered]
Background: 
- I have a collection of Customers. A customer has many properties, one of them is RevenueData.
  - RevenueData contains a List<YearlyRevenue> which in turn has a List<MonthlyRevenue>
  - MonthlyReveneue contains the revenue as a simple float
- The revenue is periodically calculated from a third party source.
- I prefer to only patch the RevenueData property instead of putting the whole Customer

This is the code (rough draft): 

` 

          var customerSelectionDtos = (await GetAllCustomers()).Where(x => x.WawiId != 0).ToList();  
          var revenueDict = new Dictionary<int, List<YearlyRevenue>>();
          var startDate = new DateTime(DateTime.Now.Year - 3, 1, 1);
          var enddate = (new DateTime(DateTime.Now.Year, DateTime.Now.Month, 1)).AddDays(-1); 

          foreach (var customer in customerSelectionDtos)
          {
              var revenue = await GetMonthlySlicedRevenue(customer.WawiId, startDate, enddate);
              revenueDict.Add(customer.WawiId, revenue);
          }

          using var asyncSession = raven.OpenAsyncSession();
          var query = asyncSession.Query<Customer>().Where(x => x.WawiId != 0).Select(x => new { x.Id, x.WawiId, x.Revenue });

          await using (var streamResults = await asyncSession.Advanced.StreamAsync(query))
          {
              while (await streamResults.MoveNextAsync())
              {
                  var customer = streamResults.Current.Document;

                  if (revenueDict.TryGetValue(customer.WawiId, out var revenue))
                  {
                      if (revenue.Count == 0)
                          continue;

                      var revenueData = customer.Revenue ?? new RevenueData();

                      asyncSession.Advanced.Patch<Customer, RevenueData>(customer.Id, x => x.Revenue, revenueData);
                  }
              }
          }

          await asyncSession.SaveChangesAsync();
`

Expected behaviour:
- Fill the RevenueData property if a customer does not yet have any revenue data
- Updated the RevenueData to the new value if a customer has an "old" value

Problem:
- Once the data is set, new values don't get written to RavenDB
- Example: 
  - A customer does not have calculated revenue yet and new value is 385.239990234375 € in January. This is saved inRavenDB
  - I change the calculation method to round at the second decimal. So the new value is 385.24. ![ApplicationFrameHost_jWUFaYAAgo](https://github.com/ravendb/ravendb/assets/43036230/cee96536-e7b1-4654-845b-bd59c5e9844d)
  - In RavenDB it's still 385.239990234375 and also the traffic log in Raven studio shows 385.239990234375 
![msedge_RRMQ66chOn](https://github.com/ravendb/ravendb/assets/43036230/2a4abc11-3a53-4b37-b477-630b712ae619)

What's going on?

Thanks a lot ✌

**Comment by ayende** [comment, marked as answer]
You are running into `float` vs. `double`, mostly likely.

See:

```
float f = 385.24f;
double d = f;
Console.WriteLine(d);
```

Note that JSON (and thus RavenDB) uses [Double-precision floating-point format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format).

You are using `Single precision` instead here.

In addition, you are running through `patch`, you are invoking a script, which round trip everything through JavaScript, and we pull just `double` from there. 

Basically, that is a serialization issue with the types. Note that when this is serialized back into a float, it will go back to `.24`

Note that for money stuff, you may need to be aware of floating points accuracy issues. 

Something that you may be able to do is use the JSON Patch mode to skip some steps and keep it in float throughout.
It _may_ work, I didn't test it, see:  https://ravendb.net/docs/article-page/5.4/python/client-api/operations/patching/json-patch-syntax

**Reply by KezimaDynamicsTK** [reply]
Huu, isn't that something...never came across that in all those years...

Thank you for the quick answer.

The source database is using float, but I don't necessarily have to do so.
I changed the MonthlyRevenue value type from float to double and take care of possible conversion issues during the calculation.

Everything is behaving as expected now.


I do like the simplicity of the JSON Patch mode, though. Could come in handy down the road.
